dagger/dagger
# feat: PAT support -- private git support on `http` / `https`

Fixes #7990

This is the PR introducing PAT support auth.

## Implementation
As requested on the issue by the community, we rely on the git credential to request for authentication:
- this allows us to maintain the same auth experience as your local `git` CLI
- When needed, the engine requests the host for the `git-credential` 
- Your local gitconfig sets up the git credential auth you want

This gives userland the power to configure auth however they want, with their preferred auth tool

## How does this work
`git` introduces a standard: [git credential](https://git-scm.com/docs/git-credential) which is queryable from the `git` CLI: 

```shell
echo -e "protocol=https\nhost=github.com\npath=dagger/dagger\n" | git credential fill
```

The engine is directly querying this command on the user's host, via the CLI and parses the output of the command ; it then injects it back inside its session, safely, as a secret

## How can you test
Please break this PR as you would, especially try to see if this maps your usual auth pattern (personal and enterprise environment): 
```shell
$ cd ~/dagger
$ ./hack/dev # wait until it finishes
$ ./hack/with-dev bash # jump into the dev version shell
$ dagger version $ shall output dagger v0.13.7-...
$ dagger -m private-module functions
```

### git credential managers

There are a lot of [git credential managers](https://github.com/git-ecosystem/git-credential-manager) out there, all mapping with the `git credential` standard. They can also be chained, and this will allow you to manage the caching side of those tokens client-side

TODO: 
- [x] Finish unit and integration test
- [x] Some user testing to see any edge case I might have missed
- [ ] Document the feature

cc @nipuna-perera @sheldonhull
